source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.cc

```diff
@@ -32,13 +32,28 @@ Http::FilterHeadersStatus AdaptiveConcurrencyFilter::decodeHeaders(Http::HeaderM
     return Http::FilterHeadersStatus::StopIteration;
   }
 
-  rq_start_time_ = config_->timeSource().monotonicTime();
+  // When the deferred_sample_task_ object is destroyed, the time difference between its destruction
+  // and the request start time is measured as the request latency. This value is sampled by the
+  // concurrency controller either when encoding is complete or during destruction of this filter
+  // object.
+  deferred_sample_task_ =
+      std::make_unique<Cleanup>([this, rq_start_time = config_->timeSource().monotonicTime()]() {
+        const auto now = config_->timeSource().monotonicTime();
+        const std::chrono::nanoseconds rq_latency = now - rq_start_time;
+        controller_->recordLatencySample(rq_latency);
+      });
+
   return Http::FilterHeadersStatus::Continue;
 }
 
 void AdaptiveConcurrencyFilter::encodeComplete() {
-  const auto rq_latency = config_->timeSource().monotonicTime() - rq_start_time_;
-  controller_->recordLatencySample(rq_latency);
+  ASSERT(deferred_sample_task_);
+  deferred_sample_task_.reset();
+}
+
+void AdaptiveConcurrencyFilter::onDestroy() {
+  deferred_sample_task_->cancel();
```

Comment:
Unfortunately this is still not right. `onDestroy()` is called in all cases, reset or otherwise, so you need to handle the case in which you have finished the encode. Please make sure you have tests that cover all these cases.

Also, I still think this pattern is a bit error prone and you would be better off just returning an `ActiveLatencySample` interface or something like that from the controller, which has a method to store the latency, and on destruction will decrement the count. I can discuss this in person tomorrow if what I'm suggesting is confusing.